src/main/java/org/terasology/metalrenegades/minimap/CharacterOverlay.java

```diff
@@ -0,0 +1,155 @@
+import org.terasology.math.geom.Rect2i;
+import org.terasology.math.geom.Vector2f;
+import org.terasology.metalrenegades.ai.component.CitizenComponent;
+import org.terasology.minimap.overlays.MinimapOverlay;
+import org.terasology.rendering.assets.texture.Texture;
+import org.terasology.rendering.nui.Canvas;
+import org.terasology.utilities.Assets;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Optional;
+
+/**
+ * This class is used to add character overlays to the minimap based on the citizen type
+ * and position
+ */
+
+
+public class CharacterOverlay implements MinimapOverlay {
+    private static final float ICON_SIZE = 16f;
+
+    private Vector2f iconSize = new Vector2f(ICON_SIZE, ICON_SIZE);
+
+    private Logger logger = LoggerFactory.getLogger(CharacterOverlay.class);
+
+    private ArrayList<EntityRef> Citizens;
+
+
+    /**
+     * This constructor sets creates the list of citizens
+     *
+     */
+
+
+    public CharacterOverlay(){
+        this.Citizens = new ArrayList<EntityRef>();
+    }
+
+    @Override
+    public void render(Canvas canvas, Rect2f worldRect) {
+
+        Collection<EntityRef> citizens = this.Citizens;
+
+
+        for (EntityRef CitizenEntity : citizens) {
+            if (!CitizenEntity.hasComponent(CitizenComponent.class)) {
+                logger.error("No Citizen found!");
+                return;
+            }
+
+            Rect2f screenRect = Rect2f.createFromMinAndSize(
+                    new Vector2f(canvas.getRegion().minX(), canvas.getRegion().minY()),
+                    new Vector2f(canvas.getRegion().maxX(), canvas.getRegion().maxY())
+            );
+
+            Rect2fTransformer transformer = new Rect2fTransformer(worldRect, screenRect);
+
+
+            LocationComponent locationComponent = CitizenEntity.getComponent(LocationComponent.class);
+            if (locationComponent == null) {
+                logger.error("Cannot find location component for Citizen: ");
+                return;
+            }
+
+            Vector2f location = new Vector2f(locationComponent.getLocalPosition().x(), locationComponent.getLocalPosition().z());
+            Vector2f mapPoint = new Vector2f(
+                    transformer.applyX(location.x),
+                    transformer.applyY(location.y)
+            );
+            Vector2f iconCenter = new Vector2f(mapPoint.x - iconSize.x / 2, mapPoint.y - iconSize.y / 2);
+
+            if (isInside(iconCenter, screenRect)) {
+                Rect2i region = Rect2i.createFromMinAndSize((int) iconCenter.x, (int) iconCenter.y, (int) iconSize.x, (int) iconSize.y);
+                String citizenType = CitizenEntity.getParentPrefab().getName();
+                Optional<Texture> icon;
+                if (citizenType.equals("MetalRenegades:marketCitizen")) {
+                    icon = Assets.getTexture("MetalRenegades:marketGooey");
+                } else {
+
+                    icon = Assets.getTexture(citizenType);
```

Comment:
This `Assets.getTexture()` will be called for every citizen, every update. So it is better if you load and store the icons beforehand in the constructor.